humanoid_league_team_communication/src/humanoid_league_team_communication/humanoid_league_team_communication.py

```diff
@@ -0,0 +1,398 @@
+            message.target_pose.position.x = self.move_base_goal.pose.position.x
+            message.target_pose.position.y = self.move_base_goal.pose.position.y
+            q = self.move_base_goal.pose.orientation
+            message.target_pose.position.z = transforms3d.euler.quat2euler([q.w, q.x, q.y, q.z])[2]
+
+        if self.ball and rospy.Time.now() - self.ball.header.stamp < rospy.Duration(self.config['lifetime']):
+            message.ball.position.x = self.ball.point.x
+            message.ball.position.y = self.ball.point.y
+            message.ball.position.z = self.ball.point.z
+
+            message.ball_confidence = self.ball_confidence
+        else:
+            message.ball_confidence = 0
+
+        if self.obstacles and rospy.Time.now() - self.obstacles.header.stamp < rospy.Duration(self.config['lifetime']):
+            for obstacle in self.obstacles.obstacles:  # type: ObstacleRelative
+                if obstacle.type in (ObstacleRelative.ROBOT_CYAN,
+                                     ObstacleRelative.ROBOT_MAGENTA,
+                                     ObstacleRelative.ROBOT_UNDEFINED):
+                    robot = robocup_extension_pb2.Robot()
+                    robot.player_id = obstacle.playerNumber
+                    robot.position.x = obstacle.pose.pose.pose.position.x
+                    robot.position.y = obstacle.pose.pose.pose.position.y
+                    q = obstacle.pose.pose.pose.orientation
+                    robot.position.z = transforms3d.euler.quat2euler([q.w, q.x, q.y, q.z])[2]
+                    team_mapping = dict(((b, a) for a, b in self.team_mapping))
+                    robot.team = team_mapping[obstacle.type]
+                    message.others.append(robot)
+                    message.obstacle_confidence.append(obstacle.pose.confidence)
+
+        if (self.ball and rospy.Time.now() - self.ball.header.stamp < rospy.Duration(self.config['lifetime']) and
+                self.pose and rospy.Time.now() - self.pose.header.stamp < rospy.Duration(self.config['lifetime'])):
+            ball_distance = math.sqrt((self.ball.point.x - self.pose.pose.pose.position.x) ** 2 +
+                                      (self.ball.point.y - self.pose.pose.pose.position.y) ** 2)
+            message.time_to_ball = ball_distance / self.config['avg_walking_speed']
+
+        if self.strategy and rospy.Time.now() - self.strategy_time < rospy.Duration(self.config['lifetime']):
+            role_mapping = dict(((b, a) for a, b in self.role_mapping))
+            message.role = role_mapping[self.strategy.role]
+
+            action_mapping = dict(((b, a) for a, b in self.action_mapping))
+            message.action = action_mapping[self.strategy.action]
+
+            side_mapping = dict(((b, a) for a, b in self.side_mapping))
+            message.offensive_side = side_mapping[self.strategy.offensive_side]
+
+        msg = message.SerializeToString()
```

Comment:
I don't know much about protobuff, but I thought it was a binary format? Why is it serialized to a string. Is it because of the old python string being basically a byte sequenceß

Reply:
Yes, serializeToString returns bytes. That's just how it works in Protobuf, might be because of the missing distinction between bytes and strings in python 2